From a8a4753329be2c37214ea496a93b5ae40b5aa5c7 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 02 Jun 2019 11:40:22 +0300
Subject: [PATCH] [SW-4185] gaudi: pll: add hbw clk sw glitch w/a

Since the clock switch has a glitch while switching between clocks
different by a factor of about 10, we W/A the problem by performing
the switch in a "gated" manner. That is, we gate the clock output,
perform the switch, and then ungate the clock output.
This way we are able to W/A the clock switch glitch as we wait enough
time for the clock to stabilize before ungating.

Change-Id: I1e455df5e923d549b65ce8ee74884da26dee430a
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/common/pll.c b/common/pll.c
index 81f2065..3e96b9d 100644
--- a/common/pll.c
+++ b/common/pll.c
@@ -25,7 +25,7 @@
 	return regmap_readl(pll_base + HL_PLL_DIV_SEL + offset);
 }
 
-static inline void hl_pll_udelay(unsigned int usec)
+void hl_pll_udelay(unsigned int usec)
 {
 	u64 cnt = (cycles_per_usec * usec) / MIN_CYCLES_PER_ITER;
 
diff --git a/common/pll.h b/common/pll.h
index 6cb640e..e788edb 100644
--- a/common/pll.h
+++ b/common/pll.h
@@ -66,5 +66,7 @@
 u64 hl_pll_get_cpu_cycles_per_usec(void);
 u32 hl_pll_get_freq(u64 pll_base, enum hl_pll_div_out div);
 bool hl_pll_is_configurable(struct hl_pll_cfg *cfg);
+void hl_pll_udelay(unsigned int usec);
+void hl_pll_wa(void);
 
 #endif /* _HL_PLL_H_ */
diff --git a/common/preboot.c b/common/preboot.c
index d476346..34c5925 100644
--- a/common/preboot.c
+++ b/common/preboot.c
@@ -175,6 +175,7 @@
 	preboot_pcie_init();
 	preboot_delay();
 	hl_pll_init();
+	hl_pll_wa();
 	/* reinit uart after PLLs configured */
 	hl_uart_init();
 	preboot_copy_spl();
diff --git a/gaudi/asic.c b/gaudi/asic.c
index 13980ad..f0d025d 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -20,7 +20,7 @@
 		.name = "DMA_W", .base = mmDMA_W_PLL_BASE,
 		.nr = 0, .nf = 59, .od = 1, .nb = 29, /* 1500 */
 		.div_factor = { 0, 1, -1, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -29,7 +29,7 @@
 		.name = "IF_W", .base = mmIF_W_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -38,7 +38,7 @@
 		.name = "IF_E", .base = mmIF_E_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -47,7 +47,7 @@
 		.name = "DMA_E", .base = mmDMA_E_PLL_BASE,
 		.nr = 0, .nf = 59, .od = 1, .nb = 29, /* 1500 */
 		.div_factor = { 0, 1, -1, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -56,7 +56,7 @@
 		.name = "MESH_W", .base = mmMESH_W_PLL_BASE,
 		.nr = 0, .nf = 59, .od = 1, .nb = 29, /* 1500 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -65,7 +65,7 @@
 		.name = "SRAM_W", .base = mmSRAM_W_PLL_BASE,
 		.nr = 0, .nf = 35, .od = 1, .nb = 17, /* 900 */
 		.div_factor = { 0, -1, -1, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -74,7 +74,7 @@
 		.name = "MESH_E", .base = mmMESH_E_PLL_BASE,
 		.nr = 0, .nf = 59, .od = 1, .nb = 29, /* 1500 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -83,7 +83,7 @@
 		.name = "SRAM_E", .base = mmSRAM_E_PLL_BASE,
 		.nr = 0, .nf = 35, .od = 1, .nb = 17, /* 900 */
 		.div_factor = { 0, -1, -1, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -92,7 +92,7 @@
 		.name = "MME", .base = mmPSOC_MME_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -111,7 +111,7 @@
 		.name = "TPC", .base = mmPSOC_TPC_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -129,7 +129,7 @@
 		.name = "NIC0", .base = mmNIC0_PLL_BASE,
 		.nr = 24, .nf = 1665, .od = 3, .nb = 832, /* 833 MHz */
 		.div_factor = { 0, 4, 8, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -138,7 +138,7 @@
 		.name = "NIC1", .base = mmNIC1_PLL_BASE,
 		.nr = 24, .nf = 1665, .od = 3, .nb = 832, /* 833 MHz */
 		.div_factor = { 0, 4, 8, -1 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -156,7 +156,7 @@
 		.name = "NIC2_MME", .base = mmNIC2_MME_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -165,7 +165,7 @@
 		.name = "NIC2_TPC", .base = mmNIC2_TPC_PLL_BASE,
 		.nr = 0, .nf = 63, .od = 1, .nb = 31, /* 1600 */
 		.div_factor = { 0, 1, 2, 7 },
-		.div_sel = { HL_PLL_PLL_CLK,
+		.div_sel = { HL_PLL_WA_INIT,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK,
 			     HL_PLL_REF_CLK
@@ -229,3 +229,34 @@
 	return factor > 2 ? factor : 2;
 }
 
+#define HL_PLL_GATER_OFFSET	0x300
+#define HL_PLL_DIVSEL_OFFSET	0x280
+
+static void hl_pll_hbw_gated_clk_sw(struct hl_pll_cfg *cfg)
+{
+	regmap_writel(0xE, cfg->base + HL_PLL_GATER_OFFSET);
+	hl_pll_udelay(1);
+	regmap_writel(HL_PLL_PLL_CLK, cfg->base + HL_PLL_DIVSEL_OFFSET);
+	hl_pll_udelay(10);
+	regmap_writel(0xF, cfg->base + HL_PLL_GATER_OFFSET);
+	hl_pll_udelay(10);
+}
+
+void hl_pll_wa(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(gaudi_pll_cfg); i++) {
+		switch (gaudi_pll_cfg[i].base) {
+		case mmPSOC_CPU_PLL_BASE:
+		case mmPSOC_PCI_PLL_BASE:
+		case mmPSOC_HBM_PLL_BASE:
+		case mmNIC2_HBM_PLL_BASE:
+			continue;
+		default:
+			break;
+		}
+
+		hl_pll_hbw_gated_clk_sw(&gaudi_pll_cfg[i]);
+	}
+}
diff --git a/gaudi/onix.c b/gaudi/onix.c
index f3e924b..60969f0 100644
--- a/gaudi/onix.c
+++ b/gaudi/onix.c
@@ -47,3 +47,5 @@
 	/* baud rate doesn't support odd number */
 	return (factor + 1) & 0xFFFE;
 }
+
+void hl_pll_wa(void) {}
diff --git a/gaudi/pldm.c b/gaudi/pldm.c
index c07682b..044289c 100644
--- a/gaudi/pldm.c
+++ b/gaudi/pldm.c
@@ -47,3 +47,5 @@
 	/* baud rate doesn't support odd number */
 	return (factor + 1) & 0xFFFE;
 }
+
+void hl_pll_wa(void) {}
diff --git a/goya/common.c b/goya/common.c
index 0dce834..ef65759 100644
--- a/goya/common.c
+++ b/goya/common.c
@@ -170,3 +170,5 @@
 {
 	return mmPSOC_GLOBAL_CONF_NON_RST_FLOPS_3;
 }
+
+void hl_pll_wa(void) {}
